### PR TITLE
Introduce Tensor overload to linspace and logspace

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105451
* __->__ #104889
* #107958

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Seems related: https://github.com/pytorch/pytorch/issues/102208

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3263,16 +3263,24 @@
 - func: ldexp.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
   tags: pointwise
 
-- func: linspace(Scalar start, Scalar end, int steps, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+- func: linspace.Scalar(Scalar start, Scalar end, int steps, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:3266]: Can you change this to just "linspace" to avoid the BC problem? native_functions.yaml should be able to handle this situation (where one overload is the "default" overload)

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -3267,12 +3267,20 @@
   dispatch:
     CompositeExplicitAutograd: linspace
 
+- func: linspace.Tensor(Tensor start, Tensor end, int steps, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+  category_override: factory
+  dispatch:
+    CompositeExplicitAutograd: linspace
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:3273]: We're going to need to add a meta registration for linspace. We should add one in https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py.

Concretely, the problem we're facing right now is that we've marked linspace.Tensor as CompositeExplicitAutograd. This automatically gives linspace.Tensor a meta registration that runs `linspace`.

Instead of doing that, we just want the meta registration to return torch.empty(steps, dtype, layout, device, pin_memory) (or steps+1? I'm not sure).

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -3267,12 +3267,20 @@
   dispatch:
     CompositeExplicitAutograd: linspace
 
+- func: linspace.Tensor(Tensor start, Tensor end, int steps, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:3270]: Before this change, the following seemed to work:
```
import torch
torch.linspace(torch.tensor(0), 1, 10)
```
After this change, it raises an error. We probably just need more overloads. The other overloads should be CompositeImplicitAutograd and just wrap the start/end tensor in a Tensor and call linspace.Tensor.

Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:3270]: Also, this change is technically BC-breaking. Previously, a model calling `torch.linspace(torch.tensor(0), torch.tensor(10), 10)` would go through the linspace.Scalar overload. Now, it goes through the linspace.Tensor overload.

I hope this doesn't actually matter in practice, but we have a FC/BC guarantee of around two weeks for our aten operators. Our tests (OSS, and then internal) will tell us if everything is OK. If they're not OK, then we're going to need to do the following:
- add linspace.Tensor, but lower its priority below linspace.Scalar so that linspace.Scalar gets triggered.
- wait two weeks
- swap the priority between linspace.Tensor and linspace.Scalar.

TL;DR: it might take some time to actually merge this PR, but we'll get there if we persevere.

Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:3270]: Edit: It isn't BC-breaking, the right word is not forwards compatible (FC):
- if a producer uses this PR and a consumer uses an older version of PyTorch, then the consumer won't have the new operator. This is an FC problem
- If a producer uses an older version of PyTorch and a consumer uses this PR, we're good, so there are no BC concerns

Given there's not really a good way to test this I think we are just going to yolo merge and see if anyone reverts.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 4]
```diff
@@ -706,6 +706,19 @@ Tensor linspace(
...
+    const Tensor& start,
+    const Tensor& end,
+    int64_t steps,
+    c10::optional<ScalarType> dtype,
+    c10::optional<Layout> layout,
+    c10::optional<Device> device,
+    c10::optional<bool> pin_memory) {
+  TORCH_CHECK(start.dim() == 0 && end.dim() == 0, "linspace only supports 0-dimensional start and end tensors, "
+    "but got start with ", start.dim(), " dimension(s) and end with ", end.dim()," dimension(s).");
+  return at::linspace(start.item(), end.item(), steps, dtype, layout, device, pin_memory);
```
Comment (zou3519) [inline aten/src/ATen/native/TensorFactories.cpp:719]: Can you add a quick test somewhere on what happens if `start` or `end` has requires_grad=True? The output of linspace should not be a part of the computation graph and should not require grad.

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -999,12 +1008,31 @@ def sample_inputs_linspace(op, device, dtype, requires_grad, **kwargs):
...
+                nstep
+            ),
+            kwargs=tensor_options
+        )
 
     yield SampleInput(1, args=(3, 1))
 
 
-def sample_inputs_logpace(op, device, dtype, requires_grad, **kwargs):
+def sample_inputs_logspace(op, device, dtype, requires_grad, **kwargs):
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:1035]: Could we add new and separate OpInfos for linspace and logspace with Tensor inputs? This will make it easier to tell if we have regressed something in the existing testing

File: torch/_refs/__init__.py [hunk 6]
```diff
@@ -4616,15 +4631,27 @@ def logspace(
...
 
     assert not isinstance(base, complex)  # for mypy
     if base < 0:
         raise NotImplementedError
-    ret = torch.linspace(
-        start,
-        end,
+    ret = torch.linspace(  # type: ignore[misc]
+        start,  # type: ignore[arg-type]
+        end,  # type: ignore[arg-type]
```
Comment (zou3519) [inline torch/_refs/__init__.py:4654]: What's going on with the typing?

File: torch/_meta_registrations.py [hunk 7]
```diff
@@ -69,6 +69,80 @@ def check_inplace_broadcast(self_shape, *args_shape):
...
+        layout=layout,
+        device="meta",
+        pin_memory=pin_memory,
+        requires_grad=requires_grad,
+    )
+
+
+@register_meta([aten.allclose.default])
+def allclose(a, b, rtol: float = 1e-05, atol: float = 1e-08, equal_nan: bool = False):
+    a.new_empty((), dtype=torch.bool)
```
Comment (zou3519) [inline torch/_meta_registrations.py:143]: allclose cannot have a meta because it returns a boolean and there is no mechanism to make the boolean symbolic with meta tensors. What are we trying to do here?

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (qqaatw) [reply aten/src/ATen/native/native_functions.yaml:3270]: Just confirmed that after the change it still works.

File: aten/src/ATen/native/TensorFactories.cpp (hunk 4 above)
Comment (qqaatw) [reply aten/src/ATen/native/TensorFactories.cpp:719]: They are end up converted to scalars before passed into the kernel so whether they require grad doesn't matter I think.

Or, do you mean adding an assertion like `TORCH_CHECK(!start.requires_grad() && !end.requires_grad(), "linspace(): The start and end tensors should not require grad.");`?

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (qqaatw) [reply torch/testing/_internal/common_methods_invocations.py:1035]: This sounds a bit weird. Do you mean that we add a new `OpInfo` identical to the current one except the `sample_inputs_*space` function is different (for testing the Tensor inputs)?

Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py:1035]: Yes

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:3270]: Sounds good

File: aten/src/ATen/native/TensorFactories.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorFactories.cpp:719]: Oh no, what I meant was a test to check if the inputs of linspace have requires_grad=True, then the output of linspace does not have requires_grad=True

File: test/functorch/test_vmap.py [hunk 8]
```diff
@@ -3733,6 +3733,8 @@ def test_vmap_exhaustive(self, device, dtype, op):
         xfail('as_strided_scatter', ''),
         xfail('equal', ''),
         xfail('linalg.lu', ''),
+        xfail('linspace', ''),
+        xfail('logspace', ''),
```
[the remark this answers is not in the document]
Comment (zou3519) [reply test/functorch/test_vmap.py:3737]: Gotcha, thanks for the clarification

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (qqaatw) [reply torch/testing/_internal/common_methods_invocations.py:1035]: If we add a new `OpInfo` with the same tested op the following line complains:

https://github.com/pytorch/pytorch/blob/35cca799ff42182a1b7f1ee4d0225ee879b7c924/torch/testing/_internal/common_device_type.py#L435

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (qqaatw) [reply aten/src/ATen/native/native_functions.yaml]: I end up added multiple overloads for different Tensor/Scalar input combinations because without these overload s some code complains that 'item' cannot be called inside vmap, for example.

OpInfo tests added.

File: aten/src/ATen/native/TensorFactories.cpp (hunk 4 above)
Comment (qqaatw) [reply aten/src/ATen/native/TensorFactories.cpp:719]: Added test in `test/test_torch.py`

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:1026]: You can have two OpInfos for the same function, you just need to add `variant_test_name` so the generated tests have different names.

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -5950,8 +5950,8 @@ def merge_dicts(*dicts):
 From PyTorch 1.11 linspace requires the steps argument. Use steps=100 to restore the previous behavior.
 
 Args:
-    start (float): the starting value for the set of points
-    end (float): the ending value for the set of points
+    start (float or Tensor): the starting value for the set of points
```
Comment (lezcano) [inline torch/_torch_docs.py:5953]: Please state that the tensor must be 0D

File: torch/_meta_registrations.py [hunk 10]
```diff
@@ -69,6 +69,83 @@ def check_inplace_broadcast(self_shape, *args_shape):
...
+        aten.logspace.default,
+        aten.logspace.Tensor_Tensor,
+        aten.logspace.Tensor_Scalar,
+        aten.logspace.Scalar_Tensor,
+        aten.logspace.Tensor_Tensor_out,
+        aten.logspace.Tensor_Scalar_out,
+        aten.logspace.Scalar_Tensor_out,
+        aten.logspace.out,
+    ]
+)
```
Comment (lezcano) [inline torch/_meta_registrations.py:91]:
```suggestion
@register_meta(
    [
        aten.linspace,
        aten.logspace
    ]
)
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (qqaatw) [reply torch/testing/_internal/common_methods_invocations.py:1040]: @peterbell10 thanks, done.

**Review by zou3519 — COMMENTED** [review]
The code looks good to me. The thing I need to back and think about is if this PR regresses any existing behavior. One way to figure this out would be to restructure the stack so that the first PR in the stack is adding the tests for Tensor `start` and `end` and the second PR is actually adding the new Tensor overloads

File: test/inductor/test_torchinductor_opinfo.py [hunk 11]
```diff
@@ -209,6 +209,10 @@ def format_op(op):
     "geometric": {f16},
     "linalg.eigh": {f32, f64},
     "linalg.eigvalsh": {f32, f64},
+    # Unsupported: data dependent operator: aten._local_scalar_dense.default
+    ("linspace", "tensor_overload"): {b8, f16, f32, f64, i32, i64},
+    # Unsupported: data dependent operator: aten._local_scalar_dense.default
+    ("logspace", "tensor_overload"): {b8, f16, f32, f64, i32, i64},
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply test/inductor/test_torchinductor_opinfo.py:215]: Restructured the stack. For the inductor specific tests, the new overloads seem to not regress the existing behavior since we xfail the tests here and the [tests](https://github.com/pytorch/pytorch/actions/runs/5922762848/job/16057727407) in the OpInfo adding PR are passing, indicating that the failures were correctly xfailed.

```
2023-08-25T17:57:21.1409490Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_linspace_tensor_overload_cuda_float16 PASSED [0.0353s] [ 41%]
2023-08-25T17:57:21.1409786Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_linspace_tensor_overload_cuda_float32 PASSED [0.0341s] [ 41%]
2023-08-25T17:57:21.1410088Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_linspace_tensor_overload_cuda_float64 PASSED [0.0354s] [ 41%]
2023-08-25T17:57:21.1410390Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_linspace_tensor_overload_cuda_int32 PASSED [0.0381s] [ 41%]
2023-08-25T17:57:21.1410688Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_linspace_tensor_overload_cuda_int64 PASSED [0.0384s] [ 41%]
```

```
2023-08-25T17:57:21.1434002Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_logspace_tensor_overload_cuda_float16 PASSED [0.0376s] [ 44%]
2023-08-25T17:57:21.1434346Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_logspace_tensor_overload_cuda_float32 PASSED [0.0360s] [ 44%]
2023-08-25T17:57:21.1434682Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_logspace_tensor_overload_cuda_float64 PASSED [0.0366s] [ 44%]
2023-08-25T17:57:21.1435112Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_logspace_tensor_overload_cuda_int32 PASSED [0.0402s] [ 44%]
2023-08-25T17:57:21.1435409Z inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_logspace_tensor_overload_cuda_int64 PASSED [0.0366s] [ 44%]
```

[the remark this answers is not in the document]
Comment (qqaatw) [reply test/inductor/test_torchinductor_opinfo.py]: Resolved :)

**Review by zou3519 — APPROVED** [review]
thanks for splitting out the OpInfos!
